samples/subsys/usb/hid-mouse/src/main.c

```diff
@@ -1,245 +1,117 @@
 			usb_wakeup_request();
 			return;
 		}
 	}
-
-	ret = gpio_pin_get(gpio, sw0.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw0.pin, ret);
-		return;
-	}
-
-	if (def_val[0] != (uint8_t)ret) {
-		state |= MOUSE_BTN_LEFT;
-	} else {
-		state &= ~MOUSE_BTN_LEFT;
-	}
-
-	if (status[MOUSE_BTN_REPORT_POS] != state) {
-		status[MOUSE_BTN_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
 }
 
-static void right_button(const struct device *gpio, struct gpio_callback *cb,
-			 uint32_t pins)
+static void input_cb(struct input_event *evt)
 {
-	int ret;
-	uint8_t state = status[MOUSE_BTN_REPORT_POS];
-
-	if (IS_ENABLED(CONFIG_USB_DEVICE_REMOTE_WAKEUP)) {
-		if (usb_status == USB_DC_SUSPEND) {
-			usb_wakeup_request();
-			return;
+	enum mouse_report_idx idx;
+	uint8_t value;
+
+	switch (evt->code) {
+	case INPUT_KEY_0:
+		rwup_if_suspended();
+		idx = MOUSE_BTN_REPORT_IDX;
+		value = report[idx];
+
+		if (evt->value) {
+			value |= MOUSE_BTN_LEFT;
+		} else {
+			value &= ~MOUSE_BTN_LEFT;
 		}
-	}
-
-	ret = gpio_pin_get(gpio, sw1.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw1.pin, ret);
-		return;
-	}
-
-	if (def_val[1] != (uint8_t)ret) {
-		state |= MOUSE_BTN_RIGHT;
-	} else {
-		state &= ~MOUSE_BTN_RIGHT;
-	}
 
-	if (status[MOUSE_BTN_REPORT_POS] != state) {
-		status[MOUSE_BTN_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
-}
+		break;
+	case INPUT_KEY_1:
+		rwup_if_suspended();
+		idx = MOUSE_BTN_REPORT_IDX;
+		value = report[idx];
 
-static void x_move(const struct device *gpio, struct gpio_callback *cb,
-		   uint32_t pins)
-{
-	int ret;
-	uint8_t state = status[MOUSE_X_REPORT_POS];
+		if (evt->value) {
+			value |= MOUSE_BTN_RIGHT;
+		} else {
+			value &= ~MOUSE_BTN_RIGHT;
+		}
 
-	ret = gpio_pin_get(gpio, sw2.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw2.pin, ret);
-		return;
-	}
+		break;
+	case INPUT_KEY_2:
+		idx = MOUSE_X_REPORT_IDX;
+		value = report[idx];
 
-	if (def_val[2] != (uint8_t)ret) {
-		state += 10U;
-	}
+		if (evt->value) {
+			value += 10U;
+		}
 
-	if (status[MOUSE_X_REPORT_POS] != state) {
-		status[MOUSE_X_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
-}
+		break;
+	case INPUT_KEY_3:
+		idx = MOUSE_Y_REPORT_IDX;
+		value = report[idx];
 
-static void y_move(const struct device *gpio, struct gpio_callback *cb,
-		   uint32_t pins)
-{
-	int ret;
-	uint8_t state = status[MOUSE_Y_REPORT_POS];
+		if (evt->value) {
+			value += 10U;
+		}
 
-	ret = gpio_pin_get(gpio, sw3.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw3.pin, ret);
+		break;
+	default:
 		return;
 	}
 
-	if (def_val[3] != (uint8_t)ret) {
-		state += 10U;
-	}
-
-	if (status[MOUSE_Y_REPORT_POS] != state) {
-		status[MOUSE_Y_REPORT_POS] = state;
-		k_sem_give(&sem);
+	if (report[idx] != value) {
+		report[idx] = value;
```

Comment:
How about memcpy'ing a local copy of the report at the function start and memcmp compare it at the end? Would save the `idx` and `value` variables, you could just `WRITE_BIT` or set the right `report` byte, would make the various `case` more compact.

Reply:
It increased flash consumption by 64 bytes.

Reply:
Just 996472 to go before running out flash! 😱